dags/tpu_observability/utils/jobset_util.py

```diff
@@ -0,0 +1,407 @@
+            template:
+              spec:
+                backoffLimit: $backoff_limit
+                completions: $completions
+                parallelism: $parallelism
+                template:
+                  spec:
+                    nodeSelector:
+                      cloud.google.com/gke-tpu-accelerator: $tpu_accelerator_type
+                      cloud.google.com/gke-tpu-topology: $tpu_topology
+                    containers:
+                    - name: $container_name
+                      image: $image
+                      command: $command
+                      args:
+                        - $args
+                      stdin: true
+                      tty: true
+                      resources:
+                        requests:
+                          google.com/tpu: $tpu_cores_per_pod
+                        limits:
+                          google.com/tpu: $tpu_cores_per_pod
+        """
+    )
+)
+
+
+@dataclasses.dataclass
+class JobSet:
+  """Generates YAML configurations for Kubernetes JobSets.
+
+  This class helps in creating JobSet YAMLs by providing a template and allowing
+  customization of various parameters like jobset name, replicas, TPU
+  configuration, and the workload script to be executed.
+
+  Attributes:
+    jobset_name: The name of the JobSet.
+    namespace: The Kubernetes namespace for the JobSet.
+    max_restarts: The maximum number of restarts for the JobSet.
+    replicated_job_name: The name for the replicated Job within the JobSet.
+    replicas: The number of replicas for the replicated Job.
+    backoff_limit: The number of failed pods to tolerate before marking the
+      Job as failed.
+    completions: The number of pods that must complete successfully.
+    parallelism: The number of pods to run in parallel.
+    tpu_accelerator_type: The type of TPU accelerator (e.g.,
+      "tpu-v6e-slice").
+    tpu_topology: The TPU topology (e.g., "4x4").
+    container_name: The name of the container in the pod.
+    image: The container image to use.
+    command: The command to run in the container.
+    tpu_cores_per_pod: The number of TPU cores requested per pod.
+  """
+
+  jobset_name: str
+  namespace: str
+  max_restarts: int
+  replicated_job_name: str
+  replicas: int
+  backoff_limit: int
+  completions: int
+  parallelism: int
+  tpu_accelerator_type: str
+  tpu_topology: str
+  container_name: str
+  image: str
+  tpu_cores_per_pod: int
+
+  def generate_yaml(self, workload_script: Workload) -> str:
+    """Generates the final JobSet YAML content.
+
+    Args:
+        workload_script: A pre-formatted, JSON-escaped string from the Workload
+          class.
+
+    Returns:
+        A string containing the complete JobSet YAML.
+    """
+    params = dataclasses.asdict(self)
+    params["command"] = ["bash", "-c"]
+    params["args"] = workload_script
+
+    return _TEMPLATE.substitute(params)
+
+
+def _get_credentials_command(node_pool: node_pool.Info):
```

Comment:
```suggestion
def _get_credentials_command(node_pool: node_pool.Info) -> str:
```

ditto for other similar functions